src/openms/source/CHEMISTRY/ModifiedPeptideGenerator.cpp

```diff
@@ -265,6 +265,161 @@ namespace OpenMS
+      }
+      return;
+    }
+
+    // iterate over each residue and build compatibility mapping describing
+    // which amino acid (peptide index) is compatible with which modification
+    map<int, vector<const ResidueModification*> > mod_compatibility;
+
+    // set terminal modifications for modifications without amino acid preference
+    for (auto const& mr : var_mods.val)
+    {
+      const ResidueModification* v = mr.first;
+
+      if (v->getTermSpecificity() == ResidueModification::N_TERM)
+      {
+        if (!peptide.hasNTerminalModification())
+        {
+          mod_compatibility[N_TERM_MODIFICATION_INDEX].push_back(v);
+        }
+      }
+      else if (v->getTermSpecificity() == ResidueModification::C_TERM)
+      {
+        if (!peptide.hasCTerminalModification())
+        {
+          mod_compatibility[C_TERM_MODIFICATION_INDEX].push_back(v);
+        }
+      }
+      else if (peptide_is_n_terminal && v->getTermSpecificity() == ResidueModification::PROTEIN_N_TERM)
+      {
+        if (!peptide.hasNTerminalModification())
+        {
+          mod_compatibility[N_TERM_MODIFICATION_INDEX].push_back(v);
+        }
+      }
+      else if (peptide_is_c_terminal && v->getTermSpecificity() == ResidueModification::PROTEIN_C_TERM)
+      {
+        if (!peptide.hasCTerminalModification())
+        {
+          mod_compatibility[C_TERM_MODIFICATION_INDEX].push_back(v);
+        }
+      }
+    }
+    
+    for (auto residue_it = peptide.begin(); residue_it != peptide.end(); ++residue_it)
+    {
+      // skip already modified residues
+      if (residue_it->isModified())
+      {
+        continue;
+      }
+
+      Size residue_index = residue_it - peptide.begin();
+
+      //determine compatibility of variable modifications
+      for (auto const& mr : var_mods.val)
+      {
+        const ResidueModification* v = mr.first;
+
+        // check if amino acid match between modification and current residue
+        if (residue_it->getOneLetterCode()[0] != v->getOrigin())
+        {
+          continue;
+        }
+
+        // Term specificity is ANYWHERE on the peptide, C_TERM or N_TERM
+        // (currently no explicit support in OpenMS for protein C-term and
+        // protein N-term)
+        // TODO This is not true anymore!
+        const ResidueModification::TermSpecificity& term_spec = v->getTermSpecificity();
+        if (term_spec == ResidueModification::ANYWHERE)
+        {
+          mod_compatibility[residue_index].push_back(v);
+        }
+        // TODO think about if it really is the same case as the one above.
+        else if (term_spec == ResidueModification::C_TERM && residue_index == (peptide.size() - 1))
+        {
+          mod_compatibility[C_TERM_MODIFICATION_INDEX].push_back(v);
+        }
+        else if (term_spec == ResidueModification::N_TERM && residue_index == 0)
+        {
+          mod_compatibility[N_TERM_MODIFICATION_INDEX].push_back(v);
+        }
```

Comment:
not sure why this is also handled here e.g. if this could lead to
two modifications (e.g., "Amidated (C-term)" and "Amidated (Protein C-term)" being applied redundantly...